Math/NumberTheory/Moduli/CubicSymbol.hs

```diff
@@ -0,0 +1,132 @@
+-- it invokes cubic reciprocity by calling cubicSymbolHelper with swapped arguments.
+cubicReciprocity :: EisensteinInteger -> EisensteinInteger -> CubicSymbol
+-- Note @cubicReciprocity 0 1 = One@. It turns out it is better to adopt this convention.
+cubicReciprocity _ 1 = One
+-- Checks if first argument is zero. Note the second argument is never zero.
+cubicReciprocity 0 _ = Zero
+-- This checks if the first argument is a unit. Because it's primary,
+-- it is enough to pattern match with 1.
+cubicReciprocity 1 _ = One
+-- Otherwise, cubic reciprocity is called.
+cubicReciprocity alpha beta = cubicSymbolHelper beta alpha
+
+-- This function takes two Eisenstein intgers @alpha@ and @beta@ and returns three
+-- arguments @(gamma, delta, contribution)@. @gamma@ and @delta@ are the associated
+-- primary numbers to alpha and beta respectively. @contribution@ is a an integer
+-- defined mod 3 which measures the difference between the cubic residue of @alpha@
+-- and @beta@ with respect to the cubic residue of @gamma@ and @delta@.
+extractPrimaryContributions :: EisensteinInteger -> EisensteinInteger -> (EisensteinInteger, EisensteinInteger, CubicSymbol)
+extractPrimaryContributions alpha beta = (gamma, delta, newSymbol)
+  where
+    newSymbol = stimes contribution Omega
+    contribution = j*m - i*m -i*n
+    m :+ n = A.quot (delta - 1) 3
+    (i, gamma) = getPrimaryDecomposition alphaThreeFree
+    (_, delta) = getPrimaryDecomposition beta
+    j = wordToInteger jIntWord
+    -- This function outputs data such that
+    -- @(1 - ω)^jIntWord * alphaThreeFree = alpha@.
+    (jIntWord, alphaThreeFree) = splitOff (1 - ω) alpha
+
+-- This function takes an Eisenstein number and returns its primary decomposition @(powerUnit, factor)@
+-- That is, given @e@ coprime with 3, it returns a unique integer (mod 6) @powerUnit@ and a unique
+-- Eisenstein number @factor@ such that @(1 + ω)^powerUnit * e = 1 + 3*factor@.
+-- Note that L.findIndex cannot return Nothing. This happens only if @e@ is not
+-- coprime with 3. This cannot happen since @U.splitOff@ is called just before.
+getPrimaryDecomposition :: EisensteinInteger -> (Integer, EisensteinInteger)
+-- This is the case where a common factor between @alpha@ and @beta@ is detected.
+-- In this instance @cubicReciprocity@ will return @Zero@.
+-- Strictly speaking, this is not a primary decomposition.
+getPrimaryDecomposition 0 = (0, 0)
+getPrimaryDecomposition e = (toInteger powerUnit, factor)
+  where
+    factor = unit * e
+    unit = (1 + ω)^powerUnit
+    powerUnit = fromMaybe
+      (error "Math.NumberTheory.Moduli.CubicSymbol: primary decomposition failed.")
+      findPowerUnit
+    -- Note that the units in @ids@ are ordered in the following way:
+    -- The i^th element of @ids@ is @(1 + ω)^i@ starting from i = 0@
+    -- That is the i^th unit counting anticlockwise starting with 1.
+    findPowerUnit = elemIndex inverseRemainder ids
```

Comment:
It seems we can do this "table lookup" more explicit, pattern-matching by all possible values of `remainder`. Pattern-matching provides an elegant way to define `symb` (equal to `stimes powerUnit Omega`), and then one can choose either `e * symbolToNum symb` or `-e * symbolToNum`, depending on which of them is equivalent to 1 modulo 3.

Reply:
As I understand it, this method is not going to work. My understanding is that you want to define `factor = e * symbolToNum symb` or `factor = -e * symbolToNum symb` depending on which one is equivalent to 1 mod 3. In general, I don't think that either of them need be equivalent to 1 mod 3. For instance, suppose that I find that `(1 + ω) * e A.rem 3 = 1`. Then its associated cubic symbol is ω. Then you see that both `e * ω` and `-e * ω` cannot have the above property. The issue is that I need to know `powerUnit` modulo 6 rather that modulo 3. One solution could be to pattern match on `remainder` and define both `factor` and `symb` at the same time. Let me know what you think.

Reply:
I have taken this approach in the latest commit. I would like to make the pattern matching more readable but the program does not compile if I use ω when I list the cases. Also, it gives me a warning for redundant pattern matching which I don't agree with.